migration/idempotent/015-materialize_operations.sql

```diff
@@ -0,0 +1,167 @@
+-- returns the time stamp before which the materialization is valid (exclusive)
+-- it is important for this to be effecient: the span_start_time index guarantees
+-- effeciency in this case.
+CREATE OR REPLACE FUNCTION _ps_trace.materialize_watermark()
+RETURNS timestamptz
+STABLE
+AS
+$func$
+    SELECT parent_start_time + interval '10 minutes' --time_bucket returns the lower end of the bucket so add an interval
+    FROM _ps_trace.operation_materialization
+    ORDER BY parent_start_time DESC
+    LIMIT 1
+$func$
+LANGUAGE SQL;
+
+--This procedure updates the materialization
+CREATE OR REPLACE PROCEDURE _ps_trace.execute_materialize_operations(log_verbose boolean = false, log_timing boolean = true)
+AS $func$
+DECLARE
+   _start TIMESTAMPTZ;
+   _global_end TIMESTAMPTZ;
+   _query_start timestamptz;
+   _bucket INTERVAL = INTERVAL '10 minute';
+BEGIN
+    --one materialization run at a time (we don't want parallel execution),
+    --so lock in a self-exclusive way.
+    --This lock also protects against concurrent data modifications which
+    --may not be strictly necessary. If this becomes an issue consider
+    --using SHARE UPDATE EXCLUSIVE instead.
+    LOCK TABLE ONLY _ps_trace.operation_materialization IN SHARE ROW EXCLUSIVE MODE;
+
+    _start := _ps_trace.materialize_watermark();
+
+    IF _start IS NULL THEN
+        --no materialization exists, get the first existing bucket in the underlyong data
+        SELECT public.time_bucket(_bucket, start_time) INTO _start
+        FROM _ps_trace.span
+        ORDER BY start_time ASC
+        LIMIT 1;
+    END IF;
+
+    --we want to materialize up to 1 bucket before the not yet filled bucket
+    --this reduces the chance of data getting backfilled and the materialization
+    --having invalid results. So, if I am seeing data at 02:23 I want to materialize
+    --up until 02:10. An alternative here is to always re-materialize the head
+    --bucket twice. Consider this if the "real-time" part of the query gets too
+    --expensive.
+    SELECT public.time_bucket(_bucket, start_time) - _bucket INTO _global_end
+    FROM _ps_trace.span
+    ORDER BY start_time DESC
+    LIMIT 1;
+
+    --we materialize one bucket at a time until we are done. We do this so that
+    --we can make steady progress even if we are pretty far behind.
+    WHILE (_start + _bucket) <= _global_end LOOP
```

Comment:
I'm not certain this should be `<=` and not just `<`. Can you please elaborate on this?